src/textual/_spatial_map.py

```diff
@@ -0,0 +1,94 @@
+from __future__ import annotations
+
+from collections import defaultdict
+from itertools import product
+from typing import Generic, Iterable, TypeVar
+
+from .geometry import Region
+
+ValueType = TypeVar("ValueType")
+
+
+class SpatialMap(Generic[ValueType]):
+    """A spatial map allows for data to be associated with a rectangular regions
+    in Euclidean space, and efficiently queried.
+
+    """
+
+    def __init__(self, grid_width: int = 100, grid_height: int = 20) -> None:
+        """Create a spatial map with the given grid size.
+
+        Args:
+            grid_width: Width of a grid square.
+            grid_height: Height of a grid square.
+        """
+        self._grid_size = (grid_width, grid_height)
+        self.total_region = Region()
+        self._map: defaultdict[tuple[int, int], list[ValueType]] = defaultdict(list)
+        self._fixed: list[ValueType] = []
+
+    def _region_to_grid(self, region: Region) -> Iterable[tuple[int, int]]:
```

Comment:
Might it be worth creating a type alias for `Coordinate` or `Position` or `Coord2D` or something of the sort? I think such an alias gives _much_ more context than just `tuple[int, int]`.

Reply:
I wouldn't have bothered personally. Sometimes indirection makes things harder to read IMHO. But maybe this is not one of those times. I've added a `GridCoordinate` type alias.